### PR TITLE
client-go: set the client retry to use exponential backoff delay with jitter by default

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
The PR sets the default for client retry to use exponential backoff delay with jitter, today the default is no backoff. 

#### Which issue(s) this PR fixes:
Fixes #105521

#### Does this PR introduce a user-facing change?
```release-note
by default, for components using client-go, the rest client will retry with an exponential backoff delay 
with some jitter. In the previous version, the client by default used to wait for 1s between retries.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
I will rebase when https://github.com/kubernetes/kubernetes/pull/106240 merges

/hold

File: staging/src/k8s.io/client-go/rest/client.go [hunk 1]
```diff
@@ -37,6 +37,15 @@ const (
 	envBackoffDuration = "KUBE_CLIENT_BACKOFF_DURATION"
 )
 
+const (
+	// default configuration for exponentially backoff delay with jitter
+	// with an initial delay of 250ms, maximum delay of 2m and
+	// a maximum of 10% jitter.
+	defaultBackoffInitialDelay = 250 * time.Millisecond
+	defaultBackoffMaxDelay     = 2 * time.Minute
+	defaultBackoffMaxJitter    = 0.1
```
Comment (tkashem) [inline staging/src/k8s.io/client-go/rest/client.go:46]: I plugged in some initial values to see how ci reacts to it.

File: staging/src/k8s.io/client-go/rest/client.go [hunk 2]
```diff
@@ -37,6 +37,15 @@ const (
 	envBackoffDuration = "KUBE_CLIENT_BACKOFF_DURATION"
 )
 
+const (
+	// default configuration for exponentially backoff delay with jitter
+	// with an initial delay of 250ms, maximum delay of 2m and
+	// a maximum of 10% jitter.
+	defaultBackoffInitialDelay = 250 * time.Millisecond
+	defaultBackoffMaxDelay     = 2 * time.Minute
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/rest/client.go:45]: 2m is a huge change compared to current "no-backoff".

I'm not saying that it's necessarily wrong one - we should just be careful with that.
It requires careful release note and I think a bit more soak-testing.

We should probably hold it until 1.24 and merge early in a cycle.

Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/client.go:45]: yes, totally agree. I put it on hold with this goal in mind - we can merge it early 1.24 to give it enough soak time. I am not comfortable merging it this late for 1.23.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: Thinking about this being a big change... it affects both responses with status 429 (Too Many Requests) and any status >= 500.  Is there any way in which this would be fine for 429, but somehow not fine for 50x?   E.g. I'm thinking about the way that maxRetries defaults to 10, so we could end up waiting for quite while if all 10 retries happen. That seems fine for 429s - since in that case the server is clearly alive and waiting is the right thing to do.  But what about when the statuses are 50x?  Is waiting for 10 retries with this level of backoff good?  I don't know... am just asking.

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/rest/client.go:45]: > Thinking about this being a big change... it affects both responses with status 429 (Too Many Requests) and any status >= 500.

I didn't yet carefully review this PR, but if that's true, I agree it's a problem.
We shouldn't be retrying 5xx (and IIRC we aren't), so backoff doesn't visibly affect those.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: I think we are retrying on 50x.  The code is this:

```
func (b *URLBackoff) UpdateBackoff(actualUrl *url.URL, err error, responseCode int) {
	// range for retry counts that we store is [0,13]
	if responseCode > maxResponseCode || serverIsOverloadedSet.Has(responseCode) {
		b.Backoff.Next(b.baseUrlKey(actualUrl), b.Backoff.Clock.Now())
		return
```
(here https://github.com/kubernetes/client-go/blob/ad95a77727e65f037cc877600f4d8ea4d867c617/rest/urlbackoff.go#L86) 

and maxResponseCode is defined to be 499.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: Makes me wonder if maybe this PR _is_ the answer afterall, with some form of dynamic adjustment as per the comments https://github.com/kubernetes/kubernetes/pull/97430

Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/client.go:45]: > Thinking about this being a big change... it affects both responses with status 429 (Too Many Requests) and any status >= 500

@wojtek-t If the server wants the caller to retry the request then it sends the following response to the caller:
```
StatusCode = {429|5xx}
Header:
  retry-After: N
```
A response status code of `503` alone from the server will not trigger a retry from the client. This is documented here:
https://github.com/kubernetes/kubernetes/blob/b429c4c9e1b3004953dfa42e46ac901bdc9e4460/staging/src/k8s.io/client-go/rest/with_retry.go#L178-L189

@JohnRusk  what you are pointing to is `UpdateBackoff` updating the backoff parameters. I know this is a little bit confusing. I am going to work on a clean up PR to make it less confusing.


> I'm thinking about the way that maxRetries defaults to 10, so we could end up waiting for quite while if all 10 retries happen

Yes, I agree, in the worst case the client can end up waiting for quite a while before giving up, especially when the request does not have a context deadline. I would like to let this PR soak for a while to see how adversely it affects.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: Ah, thank you @tkashem .  I had seen that it updated the backoff params, and forgotten that the params are only _used_ if there's a Rety-After.

BTW, I've been looking at ListWatch functionality in kublet. In that specific case, there's actually [an extra layer of retries](https://github.com/kubernetes/kubernetes/blob/b429c4c9e1b3004953dfa42e46ac901bdc9e4460/staging/src/k8s.io/client-go/tools/cache/reflector.go#L218) in addition to what we see in `Request.request()`.  That has the effect that the app does do backoff for 50x errors.  But at least its exponential by default.  I'm not sure whether we want to fix that for 500x or not. I guess the idea is that the kubelet will just keep trying and trying until the API server comes back online, and so I suppose the current behaviour there might be correct (and somewhat risky to change, due to risk of impacting kubelets ability to recover).  So I'm commenting on it here mostly just for completeness.

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/rest/client.go:45]: Yes - Kubelet has an additional layer. But we should get rid of it.
Instead of adding this in the kubelet itself we should have fixed & improved error handingling in client-go.
With the changes that are happening in this area now (we may need to go a bit futher), we should be able to get rid of this custom logic in kubelet.

Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/client.go:45]: @wojtek-t @JohnRusk does kubelet have an additional retry layer on top of the reflector, can you please point me to it?

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: I actually counted about 7 different pieces of retry logic in Kubelet!   

Sorry about the slow reply, I've been on holiday and only just got back.  I'll try to share a list of them all here.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: Ah, it looks like you've seen some of them already @tkashem  , here https://github.com/kubernetes/kubernetes/pull/102107#discussion_r635449646.  The specific one I was talking about is as per the hyperlink in my earlier comment.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: FYI @tkashem and @wojtek-t , as far as I can tell, here's the complete list of all the retry logic in Kubelet.  I'm listing it here for completeness, to answer takashem's question, and to highlight the last two points to you @wojtek-t  - since they look potentially troublesome.  They're not configurable, and their inital backoff intervals are small.

Hyperlinks point to the source code.  

1. [HTTP internal retry](https://github.com/golang/net/blob/fe4d6282115f0528fd5bc04b5644ce8ba1f43a50/http2/transport.go#L501), only for selected errors (errClientConnUnusable, errClientConnGotGoAway, ErrCodeRefusedStream aka REFUSED_STREAM).  I wonder if we get errClientConnUnusable after certain other errors earlier on the connection, but apart from that I don't think this retry mechnanism is very significant to us.
2. [Request.go main backoff](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/rest/request.go#L970). This is the subject of this current PR.  It will retry all errors which either have a Retry-After sent from the server (typically 429s) OR which look like a connection reset or app server shutdown. It will never retry other errors (e.g. TLS handshake errors, which can be common on some highly loaded servers at startup). Such errors are handled only by the various hard-coded backoffs listed below, as items 4 through 7.
3. [Request.go also respects Retry-After](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/rest/request.go#L1014). In addition to the backoff value it calculates, Request.go will also sleep for the time specified in the Retry-After response header. In practice, kube apiserver always returns 1 second as the value in that header. (As discussed here https://github.com/kubernetes/kubernetes/issues/79805)
4. [ListWatch (connections)](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/tools/cache/reflector.go#L421). Hard-coded exponential backoff.  This was enhanced by PR #100878 to backoff on 429s.  That PR adds a layer of backoff for 429s over and above any done by the Request level backoff above. Although I'm not sure how necessary that will be going forward (i.e. if this current PR is approved).  Perhaps that fits with wojeck-t's comment above.
5. [ListWatch (outer loop)](https://github.com/kubernetes/client-go/blob/782ff783b635df54ddf44d55f3bd3e48eb4dcb9a/tools/cache/reflector.go#L220). This is the one I mentioned earlier tkashem.  It's a hard-coded exponential backoff.
6. [csi-plugin waitForAPIServerForEver](https://github.com/kubernetes/kubernetes/blob/f0dbc32ed9c8a51a6d149eb62d868fb995f0db3a/pkg/volume/csi/csi_plugin.go#L1053).  This appears to be hard coded a 1 second delay,with no backoff.
7. [csi-pluging main backoffs](https://github.com/kubernetes/kubernetes/blob/f0dbc32ed9c8a51a6d149eb62d868fb995f0db3a/pkg/volume/csi/csi_plugin.go#L292).  Two levels of nested exponential backoff, with hard-coded parameters.  If all requests fail, it will retry a total of 24 times with a total waiting time of 150 seconds, then restart the kubelet.   The intial backoff interval is only 15 milliseconds.  Which doesn't sound very big, especially in the context of a large cluster with thousands of kubelets.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: Two more, both with no backoff (in most cases) and hard coded-retry limit.
9. kubelet_node_statsu.go: tryUpdateNodeStatus
10. k8s.io/component-helpers/apimachinery/lease/controller.go: retryUpdateLease.

I will probably post more about these two in a new Issue, but need to investigate them a little more first.

Comment (JohnRusk) [reply staging/src/k8s.io/client-go/rest/client.go:45]: The issue with those last two turns out to be not primarily a retry issue. They are _designed_ to reset all TCP connections after only the second failure.  But that causes a problem. In fact, it turns out to be the _only_ significant retry-related problem that I've found when testing with 5000 nodes.  Fixing that one problem is necessary and sufficient to solve all the apparent retry problems in my 5000 node test clusters.    I.e. when I solved that, issue #65954 was solved on my cluster.  But wiwthout that fix, nothing else could solve #65954.

I've logged a PR here to fix it #107781.

**Comment by tkashem** [comment]
we are adding a new metric to count the retries explicitly - https://github.com/kubernetes/kubernetes/pull/108396

**Comment by tkashem** [comment]
```
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633
Mar  3 14:30:11.630: Pod didn't start within time out period
Unexpected error:
    <*errors.errorString | 0xc0002bc250>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/proxy.go:408
```

/test pull-kubernetes-e2e-kind

**Comment by tkashem** [comment]
as discussed, we will merge it early in 1.25 cycle to give it enough soak time.

**Comment by djsly** [comment]
Any plan to cherry pick this ?

**Comment by tkashem** [comment]
> Any plan to cherry pick this ?

@djsly if you are looking to enable exponential backoff in your kubernetes application then you can define these env variables https://github.com/kubernetes/kubernetes/blob/4ce5a8954017644c5420bae81d72b09b735c21f0/staging/src/k8s.io/client-go/rest/client.go#L34-L37

we will probably observe client behavior with this enabled for some time, we can potentially cherry-pick it if the results are satisfactory.

**Comment by aojea** [comment]
right now an apiserver not ready or a server resetting the connections blocks a client-go for 10 seconds by default

https://gist.github.com/aojea/31ab71c894c15f46a567c5e8aa235a17


how long it will take with this change? if is different it should be mentioned in the release notes

File: staging/src/k8s.io/client-go/rest/request.go [hunk 3]
```diff
@@ -129,7 +129,7 @@ func NewRequest(c *RESTClient) *Request {
 		backoff = c.createBackoffMgr()
 	}
 	if backoff == nil {
-		backoff = noBackoff
+		backoff = defaultBackoffManager()
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/rest/request.go:132]: If we want to start using exponential backoff, I would like this TODO to be addressed:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/with_retry.go#L225-L229

[ideally in a separate PR first, but might be separate commit here]

I would like to avoid increasing every sleep by additional 1s...

File: staging/src/k8s.io/client-go/rest/client.go [hunk 4]
```diff
@@ -37,6 +37,15 @@ const (
 	envBackoffDuration = "KUBE_CLIENT_BACKOFF_DURATION"
 )
 
+const (
+	// default configuration for exponentially backoff delay with jitter
+	// with an initial delay of 250ms, maximum delay of 1m and
+	// a maximum of 10% jitter.
+	defaultBackoffInitialDelay = 250 * time.Millisecond
+	defaultBackoffMaxDelay     = 1 * time.Minute
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/rest/client.go:45]: 1m sounds quite high to me - given we start from current 1s, I would try with something smaller first.

We generally set maxRetries to 10:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L153


Also @aojea comment:
https://github.com/kubernetes/kubernetes/pull/106272/#issuecomment-1120049668
is relevant here

I would probably decrease to sth like 10-15s for now.
[And as mentioned by @aojea - should be fixed in the release notes]

**Comment by tkashem** [comment]
@wojtek-t @aojea I also noticed that when we return the maximum delay it is not jittered, so after a few retries we keep returning a fixed wait of `20s` (max duration)
```
sleep sequence with retries: [0s 1s 1.009195954s 1s 2.080410449s 1s 4.338292259s 1s 8.774759732s 1s 17.614697733s 1s 20s 1s 20s 1s 20s 1s 20s 1s 20s]
```

So I changed the `Backoff` logic to jitter the max duration as well, please review it and let me know if this is okay to add.

File: staging/src/k8s.io/client-go/rest/request.go (hunk 3 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request.go:130]: yes, i would like to address this in a follow up PR - we already have an issue tracking this.

File: staging/src/k8s.io/client-go/rest/client.go (hunk 4 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/client.go]: okay, so i changed the initial delay to `1s` and max delay to `20s`. initial delay of `1s` to maintain current behavior - today by default we wait `1s` before retrying. 
https://github.com/kubernetes/kubernetes/blob/b53be1d66ef1c7f79410d619864d3788e084dd49/staging/src/k8s.io/client-go/rest/with_retry.go#L225-L229

I will remove the above with a follow up PR.

**Comment by tkashem** [comment]
fixing a unit test failure...

File: staging/src/k8s.io/client-go/rest/request.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/rest/request.go:130]: So my point is - we need to address that one first - otherwise this PR automatically means that in the first retry we wait 2s instead of 1s.

We have to merge the PR fixing that issue not later than this PR...

File: staging/src/k8s.io/client-go/rest/client.go [hunk 5]
```diff
@@ -37,6 +37,15 @@ const (
 	envBackoffDuration = "KUBE_CLIENT_BACKOFF_DURATION"
 )
 
+const (
+	// default configuration for exponentially backoff delay with jitter
+	// with an initial delay of 250ms, maximum delay of 1m and
+	// a maximum of 10% jitter.
+	defaultBackoffInitialDelay = 1 * time.Second
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/rest/client.go:44]: I would leave 250ms as you had.
We should just be maxing that with the 1s returned from Retry-After - which you will get if you first address the issue I mentioned below.

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go [hunk 6]
```diff
@@ -100,7 +100,10 @@ func (p *Backoff) Next(id string, eventTime time.Time) {
 	} else {
 		delay := entry.backoff * 2       // exponential
 		delay += p.jitter(entry.backoff) // add some jitter to the delay
-		entry.backoff = time.Duration(integer.Int64Min(int64(delay), int64(p.maxDuration)))
+		// instead of returning a fixed value of p.maxDuration, we can
+		// jitter the maximum duration.
+		maxDurationJittered := p.maxDuration + p.jitter(p.maxDuration)
+		entry.backoff = time.Duration(integer.Int64Min(int64(delay), int64(maxDurationJittered)))
```
Comment (aojea) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:106]: why do we need to jittered values, we already add jitter to the delay

Comment (aojea) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:106]: also, this is breaking

> // move backoff to the next mark, capping at maxDuration

no?

File: staging/src/k8s.io/client-go/rest/request.go (hunk 3 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request.go:130]: https://github.com/kubernetes/kubernetes/pull/110100

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff.go (hunk 6 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff.go]: I removed it, i think the preceding sequence of jittered delays will prevent a synchronized cluster of calls on the server.

**Comment by wojtek-t** [comment]
I'm fine with this change in the current shape.
But @aojea had some comments before so I want to give him a chance to take a look before tagging.

File: staging/src/k8s.io/client-go/rest/client.go [hunk 7]
```diff
@@ -145,12 +167,13 @@ func readExpBackoffConfig() BackoffManager {
 	backoffBaseInt, errBase := strconv.ParseInt(backoffBase, 10, 64)
 	backoffDurationInt, errDuration := strconv.ParseInt(backoffDuration, 10, 64)
 	if errBase != nil || errDuration != nil {
-		return &NoBackoff{}
+		return defaultBackoffManager()
```
Comment (aojea) [inline staging/src/k8s.io/client-go/rest/client.go:170]: is this right? 
if there is an error reading the config I think that we should not use a backoff manager instead of using the default

File: staging/src/k8s.io/client-go/rest/client.go [hunk 8]
```diff
@@ -145,12 +167,13 @@ func readExpBackoffConfig() BackoffManager {
...
 	}
 	return &URLBackoff{
-		Backoff: flowcontrol.NewBackOff(
+		Backoff: flowcontrol.NewBackOffWithJitter(
 			time.Duration(backoffBaseInt)*time.Second,
-			time.Duration(backoffDurationInt)*time.Second)}
+			time.Duration(backoffDurationInt)*time.Second,
+			defaultBackoffMaxJitter)}
```
Comment (aojea) [inline staging/src/k8s.io/client-go/rest/client.go:176]: shouldn't this be configurable too?

File: staging/src/k8s.io/client-go/rest/request_test.go [hunk 9]
```diff
@@ -2065,6 +2065,13 @@ func testRESTClientWithConfig(t testing.TB, srv *httptest.Server, contentConfig
 	if err != nil {
 		t.Fatalf("failed to create a client: %v", err)
 	}
+
+	// the tests by default can continue to use NoBackoff
+	// to maintain shorter test duration.
```
Comment (aojea) [inline staging/src/k8s.io/client-go/rest/request_test.go:2070]: is there so much difference? are we retrying on the tests?

if possible I think that is preferable to use same defaults in the tests

File: staging/src/k8s.io/client-go/rest/client.go (hunk 7 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/client.go:170]: with this PR the default is 'exponential backoff' - so if there is an error this means that we can't honor the values they specified in the env variables, instead we are going to use the default values. But they are still getting exponential backoff as they originally asked for.

File: staging/src/k8s.io/client-go/rest/client.go (hunk 8 above)
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/client.go:176]: i am not sure that it matters a lot, but i am fine making it configurable too. any thoughts here @wojtek-t?

File: staging/src/k8s.io/client-go/rest/request.go [hunk 10]
```diff
@@ -222,7 +220,7 @@ func (r *Request) Resource(resource string) *Request {
 // or defaults to the stub implementation if nil is provided
 func (r *Request) BackOff(manager BackoffManager) *Request {
 	if manager == nil {
-		r.backoff = &NoBackoff{}
+		// bad input, so we will not change the current backoff
```
[the remark this answers is not in the document]
Comment (tkashem) [reply staging/src/k8s.io/client-go/rest/request.go:223]: when a user call `request.BackOff` this implies that the user wants to specify a custom `Backoff`, and passing `nil` is a programmer error. 
Ideally I think we should set the `Request.err` to "can't set a nil Backoff" and render the request object not usable. But this will also be a breaking change since some folks may be using `request.BackOff(nil)` as a way to reset the underlying backoff manager for the request.

we have these options:
- A: do nothing, it's a programmer error (not backward compatible)
- B: set the backoff to default `r.backoff = defaultBackoffManager()` - we are making a choice for the user here when he/she asked for a different thing maybe.
- C: we can go back to `r.backoff = &NoBackoff{}`, this retains old behavior, what do you think?

[the remark this answers is not in the document]
Comment (aojea) [reply staging/src/k8s.io/client-go/rest/request.go:223]: I'm always vote for being conservative and keep compatibility unless is clear a bug, so I would say C, but let's wait for wojtek or jordan

[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/rest/request.go:223]: OK - I agree with Antonio - let's go with (C) and retain the old behavior.
It's something that we will be able to change later if needed independently, so let' try to be consiervative with this PR and we can potentially discuss that as a follow up.

File: staging/src/k8s.io/client-go/rest/client_test.go [hunk 11]
```diff
@@ -360,14 +360,14 @@ func TestCreateBackoffManager(t *testing.T) {
 		t.Errorf("Zero backoff duration, but backoff still occurring.")
 	}
 
-	// No env -> No backoff.
+	// No env -> default is exponential backoff with jitter.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/client_test.go:363]: why is it ok to change the default when no envvars are set?

File: staging/src/k8s.io/client-go/rest/request_test.go (hunk 9 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/rest/request_test.go:2070]: the fact that we had to adjust the tests to keep passing with the default changing to backoff is a strong signal that this will be a user-visible change that will disrupt/break some consumers

**Comment by liggitt** [comment]
in my opinion, adding jitter to the backoff seems more reasonable than switching the default behavior to exponential delay

**Comment by liggitt** [comment]
/hold

it's not clear to me that exponential backoff is correct... the parent issue cites clients following the `retry-after: 1` as problematic, but that seems like exactly what the client should do... follow the backoff instructions of the server.

If we want to jitter the client slightly, that could be ok. Separately, if we want clients to back off more than 1s, why aren't we doing that by modifying the server response?

**Comment by aojea** [comment]
/lgtm cancel

Jordan is right, my apologies, I've should have tested this before LGTMd

I've used these snippets for testing https://gist.github.com/aojea/31ab71c894c15f46a567c5e8aa235a17#file-results-md

For a server returning 503 and retry after with value :1 
```
I0525 16:51:50.183595 2058136 round_trippers.go:574] Response Status: 500 Internal Server Error in 0 milliseconds
----------------------------
duration error: 34.590216936s 
 err: client rate limiter Wait returned an error: context deadline exceeded
```

However, for an scenarios with connection refused errors, the behavior is the same :shrug: 

```
I0525 16:54:44.003365 2060311 round_trippers.go:473]     Accept: application/json, */*
I0525 16:54:44.003703 2060311 round_trippers.go:574] Response Status:  in 0 milliseconds
I0525 16:54:44.003724 2060311 urlbackoff.go:92] Client is returning errors: code 0, error Get "http://127.0.0.1:34689/?timeout=30s": EOF
----------------------------
duration error: 10.007487528s 
 err: Get "http://127.0.0.1:34689/?timeout=30s": EOF
Error accepting:  accept tcp 127.0.0.1:34689: use of closed network connection
exit status 1
```

I was thinking that the backoff applied to the network errors problem, that makes sense, because we don't want to bomb the server and we don't know when the connectivity will be back.

But this seems to do just the opposite, we  keep the same behavior for network errors and we don't leverage the Retry-after answer, that looks wrong to me

**Comment by aojea** [comment]
It seems that the exponential backoff can be applied only when there is no answer (network error, it seems we have an artificial retry-after with a constant https://github.com/kubernetes/kubernetes/blob/1290d6a4c0c5cc4f758b428993d2c3760d9cf279/staging/src/k8s.io/client-go/rest/with_retry.go#L353

or not "retry-after" header has been received

**Comment by wojtek-t** [comment]
> Separately, if we want clients to back off more than 1s, why aren't we doing that by modifying the server response?

I think that tracking that in the context of client will be hard on the server side (we don't want to track some per-client information on the server side).

But thinking about that, maybe we actually don't need it. We may just adjust the "retry-after" value based on the load on the apiserver not in the context of requests. This may potentially be unfair, but that may be acceptable.


Also - I thought that it's mostly for 429s (for which this change makes sense to me), but I agree that for 5xx ones it doesn't seem right...

**Comment by tkashem** [comment]
> Also - I thought that it's mostly for 429s (for which this change makes sense to me), but I agree that for 5xx ones it doesn't seem right...

I addded a unit test that documents cases on which we retry - 429, 5xx, and retryable errors. https://github.com/kubernetes/kubernetes/pull/110253

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR
